File: libs/srs/src/lib/lessons.ts
import type { Modify } from "@koloda/app";
import type { MessageDescriptor } from "@lingui/core";
import { msg } from "@lingui/core/macro";
import { z } from "zod";
import type { LessonAlgorithm } from "./algorithms";
import type { Card } from "./cards";
import type { Deck } from "./decks";
import type { InsertReviewData } from "./reviews";
import type { LessonTemplateRow, Template, TemplateField, TemplateLayoutItem } from "./templates";

export const LESSON_TYPES = ["untouched", "learn", "review", "total"] as const;

export const LESSON_TYPE_LABELS: Record<LessonType, MessageDescriptor> = {
  untouched: msg`lesson.init.labels.untouched`,
  learn: msg`lesson.init.labels.learn`,
  review: msg`lesson.init.labels.review`,
  total: msg`lesson.init.labels.total`,
} as const;

export type LessonType = (typeof LESSON_TYPES)[number];

export const lessonDeckSchema = z.object({
  // WHY: raw `db.execute` rows may surface int4/counts as string; coerce at this boundary.
  id: z.coerce.number().int(),
  title: z.string(),
  untouched: z.coerce.number(),
  learn: z.coerce.number(),
  review: z.coerce.number(),
  total: z.coerce.number(),
});

export type LessonDeck = z.infer<typeof lessonDeckSchema>;

export type LessonsResult = {
  total: LessonAmounts;
  decks: LessonDeck[];
};

export type LessonTableRow = LessonAmounts & {
  id: Deck["id"] | null;
  title: string | null;
};

export function toLessonTableRows({ total, decks }: LessonsResult): LessonTableRow[] {
  // WHY: Aggregate stays first to match the old UNION ALL … ORDER BY id NULLS FIRST shape.
  // LessonsTable pins row "0"; do not move the total row to the end.
  return [{ id: null, title: null, ...total }, ...decks];
}

export type LessonFilters = { deckIds?: Deck["id"][] };

export type LessonAmounts = Record<LessonType, number>;

export type LessonTemplateLayoutItem = Modify<
  TemplateLayoutItem,
  {
    field: TemplateField | undefined;
  }
>;

export type LessonTemplate = Modify<
  Template,
  {
    layout: LessonTemplateLayoutItem[];
  }
>;

export type GetLessonDataParams = {
  filters: LessonFilters;
  amounts: LessonAmounts;
};

export type LessonData = {
  cards: Card[];
  decks: Deck[];
  templates: LessonTemplate[];
  algorithms: LessonAlgorithm[];
};

export type LessonResultData = {
  card: Card;
  review: InsertReviewData;
};

export function convertTemplateToLessonTemplate(
  template: Pick<Template, "id" | "content"> | LessonTemplateRow,
): LessonTemplate {
  const layout: LessonTemplateLayoutItem[] = template.content.layout.map((entry) => ({
    ...entry,
    field: template.content.fields.find((x) => x.id === entry.field),
  }));
  return { ...template, layout } as LessonTemplate;
}

File: apps/native-tauri/src-tauri/src/domain/lessons.rs
use serde::{Deserialize, Serialize};

use super::templates::TemplateField;
use crate::app::utility::{default_now, serialize_optional_timestamp, serialize_timestamp};
use crate::domain::cards::{Card, UpdateCardProgress};
use crate::domain::decks::Deck;
use crate::domain::reviews::InsertReviewData;

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Lesson {
    pub id: Option<i64>,
    pub title: Option<String>,
    pub untouched: i64,
    pub learn: i64,
    pub review: i64,
    pub total: i64,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct LessonFilters {
    pub deck_ids: Option<Vec<i64>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct LessonAmounts {
    pub untouched: i64,
    pub learn: i64,
    pub review: i64,
    pub total: i64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct LessonTemplateLayoutItem {
    pub field: Option<TemplateField>,
    pub operation: String,
    pub field_id: i64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct LessonTemplate {
    pub id: i64,
    pub title: String,
    pub fields: Vec<TemplateField>,
    pub layout: Vec<LessonTemplateLayoutItem>,
    #[serde(serialize_with = "serialize_timestamp")]
    pub created_at: i64,
    #[serde(default, serialize_with = "serialize_optional_timestamp")]
    pub updated_at: Option<i64>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct LessonData {
    pub cards: Vec<Card>,
    pub decks: Vec<Deck>,
    pub templates: Vec<LessonTemplate>,
    pub algorithms: Vec<super::algorithms::Algorithm>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GetLessonDataParams {
    #[serde(default = "default_now")]
    pub due_at: i64,
    pub filters: LessonFilters,
    pub amounts: LessonAmounts,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct LessonResultData {
    pub card: UpdateCardProgress,
    pub review: InsertReviewData,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GetLessonsParams {
    pub due_at: i64,
    pub filters: Option<LessonFilters>,
}
